core/test/invariants/ERC4626Invariants.t.sol

```diff
@@ -0,0 +1,110 @@
+                lenderImplementation,
+                abi.encodePacked(address(asset))
+            ));
+            RateModel rateModel = new RateModel();
+            lender.initialize(rateModel, 8); // TODO: replace 8 with an env var
+
+            vault = ERC4626(address(lender));
+            vaultHarness = new ERC4626Harness(lender);
+
+            targetContract(address(vaultHarness));
+
+            // forge can't simulate transactions from addresses with code, so we must exclude all contracts
+            excludeSender(address(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D)); // vm
+            excludeSender(address(0x4e59b44847b379578588920cA78FbF26c0B4956C)); // built-in create2 deployer
+            excludeSender(address(this));
+            excludeSender(address(asset));
+            excludeSender(address(lenderImplementation));
+            excludeSender(address(rateModel));
+            excludeSender(address(vault));
+            excludeSender(address(vaultHarness));
+        }
+
+        thingsThatShouldntChange = ThingsThatShouldntChange(
+            vault.name(),
+            vault.symbol(),
+            vault.decimals(),
+            vault.asset(),
+            vault.DOMAIN_SEPARATOR()
+        );
+    }
+
+    function invariant_thingsThatShouldntChangeDontChange() public {
+        ThingsThatShouldntChange memory update = ThingsThatShouldntChange(
+            vault.name(),
+            vault.symbol(),
+            vault.decimals(),
+            vault.asset(),
+            vault.DOMAIN_SEPARATOR()
+        );
+
+        assertEq(bytes(update.name), bytes(thingsThatShouldntChange.name));
+        assertEq(bytes(update.symbol), bytes(thingsThatShouldntChange.symbol));
+        assertEq(update.decimals, thingsThatShouldntChange.decimals);
+        assertEq(address(update.asset), address(thingsThatShouldntChange.asset));
+        assertEq(update.domainSeparator, thingsThatShouldntChange.domainSeparator);
+    }
+
+    function invariant_conversionsAreInverses() public {
+        // Choice of `x` is random/convenient
+        uint256 x = vault.totalSupply();
+
+        // This would need to be true for all `x` in order to fully prove they're inverses
+        assertEq(vault.convertToAssets(vault.convertToShares(x)), x);
+        assertEq(vault.convertToShares(vault.convertToAssets(x)), x);
+    }
+
+    function invariant_totalSupplyEqualsSumOfBalances() public {
+        uint256 totalSupply;
+        uint256 count = vaultHarness.getHolderCount();
+        for (uint256 i = 0; i < count; i++) {
+            totalSupply += vault.balanceOf(vaultHarness.holders(i));
+        }
+        assertEq(totalSupply, vault.totalSupply());
+    }
+
+    function invariant_maxRedeemLessThanBalance() public {
+        uint256 count = vaultHarness.getHolderCount();
+        for (uint256 i = 0; i < count; i++) {
+            address user = vaultHarness.holders(i);
+
+            assertLe(vault.maxRedeem(user), vault.balanceOf(user));
```

Comment:
Nice assert  🙌  

I think this demonstrates how we're complying with certain requirements like this one:

> MUST return the maximum amount of shares that could be transferred from owner through redeem and not cause a revert, which MUST NOT be higher than the actual maximum that would be accepted (it should underestimate if necessary).

ERC4646 goes a step further to even include this requirement to indicate what should happen when limits are imposed. 

> MUST factor in both global and user-specific limits, like if redemption is entirely disabled (even temporarily) it MUST return 0.

Can limits be imposed from the contracts in Aloe II such that 0 will be returned? If so, do we need some test to demonstrate this?

Reply:
In the harness, we use `vault.maxRedeem(user)` as an upper bound on the random/fuzzed redemption amount. This implies that (sometimes), the redemption amount will exactly equal `vault.maxRedeem(user)`. The harness will then test like so:

- The value from `maxRedeem` corresponds to some amount of underlying asset. If that amount is 0, assert that it reverts with `"Aloe: zero impact"`
- Otherwise, assert successful transaction, appropriate event emission, and correct balance changes

So to answer your question: no, we don't have any kind of emergency/disable functionality right now. But if we did, I think the harness would be sufficient to check that `maxRedeem` is accurate and precise at the user level. And if it's good at the user level _for all users_, that should imply it's good globally.